master/internal/resourcemanagers/agent/agent.go

```diff
@@ -418,22 +418,39 @@ func (a *agent) handleAPIRequest(ctx *actor.Context, apiCtx echo.Context) {
 }
 
 func (a *agent) handleIncomingWSMessage(ctx *actor.Context, msg aproto.MasterMessage) {
+	log := ctx.Log().WithField("agent-id", ctx.Self().Address().Local())
 	switch {
 	case msg.AgentStarted != nil:
 		ctx.Log().Infof("agent connected ip: %v resource pool: %s slots: %d",
 			a.address, a.resourcePoolName, len(msg.AgentStarted.Devices))
 
-		// TODO(ilia): Error out on a change in devices.
-		if !a.started {
+		if a.started {
+			err := a.agentState.checkAgentStartedDevicesMatch(ctx, msg.AgentStarted)
+			if err != nil {
+				log.WithError(err).
+					Error("change in agent devices was detected")
+				wsm := ws.WriteMessage{
+					Message: aproto.AgentMessage{
+						AgentShutdown: &aproto.AgentShutdown{
+							ErrMsg: aproto.ErrAgentMustReconnect.Error(),
+						},
+					},
+				}
+				if err = ctx.Ask(a.socket, wsm).Error(); err != nil {
+					log.WithError(err).Error("failed to tell agent to reconnect")
+				}
+				check.Panic(err)
+				ctx.Self().Stop()
```

Comment:
so weird to reason about stop not actually happening until we get around to it in the inbox